---
title: "Other important scripts-plots"
author: "Ariadna Colmenero Cobo de Guzm√°n"
output: html_document
---

# 1. PREPARATION OF THE QUALITY METRICS BARPLOTS

Once we have obtained the Quality Control Metrics, we can plot some of the information that it provides to us. This will make it more informative.

```{r}
library(dplyr) # A fast, consistent tool for working with data frame like objects, both in memory and out of memory.

D <- read.csv(file.choose(), header=T, sep="\t")
barplot(D,beside=TRUE)
rownames(D) <- D$Sample.Name

D <- D[,-1]

df_max <- D  %>%
  mutate(Low_Quality=ifelse(Mean.target.coverage==min(Mean.target.coverage), "1", "0"))

layout(matrix(c(1:2), nrow=1, byrow=FALSE))

layout.show(2)

legend(x = "top", legend = colnames(D), bty = "n", ncol = 1, inset = -0.15)

barplot(t(as.matrix(D[, 1])), 
        beside = TRUE,
        names.arg = rownames(D),
        legend.text = TRUE,
        ylim = c(0,500),
        ylab = "Mean target coverage ",
        las = 3)
abline(h = c(100, 30),
       col = c("#956565", "#640B0B"),
       lwd = 2,
       lty = 2:3)

legend(x = "top", legend = colnames(D), bty = "n", ncol = 1, inset = -0.15)

barplot(t(as.matrix(D[, 2])), 
        beside = TRUE,
        names.arg = rownames(D),
        legend.text = TRUE,
        ylim = c(0,200),
        ylab = "Total million reads",
        las = 3)

abline(h = 10,
       col = "#640B0B",
       lwd = 2,
       lty = 2:3)
```
Now, it should be noted that the thresholds have been established by own criteria. Still, such graphs allow us to see, in this case, the overall poor quality of these samples.


# 2. CORRELATION AGE-COVERAGE

First, we loaded a data set containing information for the 15 WES samples including the coverage and the year in which the FFPE sample was prepared.

```{r}
library(corrplot) # Provides a visual exploratory tool on correlation matrix.
library(ggplot2) # A system for 'declaratively' creating graphics, based on "The Grammar of Graphics".
library(ggpubr) # For elegant data visualization in R.  
```

```{r}
A <- read.csv(file.choose(), header=T, sep="\t")

rownames(A) <- A$Sample.Name
A <- A[,-1]
```

Next, in order to know what type of correlation to apply, we test whether the two variables follow a normal distribution. 

```{r}
for (i in 1:2){print(shapiro.test(A[,i]))} # They both follow a normal distribution.
```
As both variables follow a normal distribution, i.e. they have a p-value above the significance level, we can apply a Pearson's correlation.

```{r}
corMatMy <- cor(A, method= "pearson")
corMatMy <- as.matrix(corMatMy)
corMatMy

corrplot(corMatMy, order = "hclust", tl.cex = 0.9, tl.col = "darkblue", mar=c(0,0,1,0), font.main = 2)
```
Another interesting question is whether these correlations are significant. To do this, we use the cor.test() function.

```{r}
Mean_target_coverage <- A$Mean.target.coverage
Age_FFPE_tissue <- A$Age.FFPE.tissue

res <- cor.test(x=Mean_target_coverage, y=Age_FFPE_tissue, method="pearson")
res$p.value
res$estimate
```
Finally, and for illustrative purposes, we draw up a scatter plot with the correlation coefficient:

```{r}
A <- A[,c(1,3)]
sp <- ggscatter(A, x = "Age.FFPE.tissue", y = "Mean.target.coverage",
                add = "reg.line",  # Add regressin line
                add.params = list(color = "blue", fill = "lightgray"), # Customize reg. line
                conf.int = TRUE, # Add confidence interval
                xlim = c(1995, 2014)
)

# Add correlation coefficient
sp + stat_cor(method = "pearson", label.x = 2000)
#> `geom_smooth()` using formula 'y ~ x'
```

# 3. BOXPLOT FFPE/TS TISSUE

The following code is then applied to determine whether there is a relationship between the type of material of origin of the sample, i.e. whether it is Frozen Tissue or FFPE.

```{r}
library(ggplot2)
library(corrplot)
```
So, first, we load the dataset with the information for these variables.

```{r}
# Data 
D <- read.csv(file.choose(), header=T, sep="\t")
D
rownames(D) <- D$Sample_ID
```
The first step, and in order to see it visually, a boxplot is represented for each type of origin of the sample tissue.

```{r}
# Boxplot:
ggplot(data = D, mapping = aes(x = Material, y = Mean_target_coverage, group= Material))+
  geom_boxplot(outlier.shape = NA,alpha=0.5)+
  # geom_violin(scale = "width",alpha=0.4)+
  geom_jitter(size=5, width = 0.2, pch=21, colour="black", aes(fill=Material))+
    theme(axis.text.x = element_text(angle = 45,vjust = 1,hjust = 1))+
  theme_bw(base_size = 11, base_family = "")
```

Next, as the samples do not follow a normal distribution, we proceed to look at the Spearman correlation. It should be noted that all these analyses would be more powerful and meaningful if we increased the number of samples, especially in the case of the FTs.

```{r}
for (i in 1:3){print(shapiro.test(D[,i]))}

corMatMy <- cor(D, method= "spearman")
corMatMy <- as.matrix(corMatMy)
corMatMy

corrplot(corMatMy, order = "hclust", tl.cex = 0.9, tl.col = "darkblue", mar=c(0,0,1,0), font.main = 2)
corrplot(corMatMy)
```

We cannot perform a t-test because the samples are not normally distributed and, moreover, they do not have a sample size greater than 30, which could justify some skewness in the normality. Therefore, we proceed with the Wilcoxon test:

```{r}
wilcox.test(Mean_target_coverage~Material,data=D)

Mean_target_coverage <- D$Mean_target_coverage
Material <- D$Material

res <- cor.test(x=Mean_target_coverage, y=Material, method="spearman")
res$p.value
res$estimate
```

# 4. Insert size metrics

First of all, in order to compare the values of insert size metrics for both groups of samples, we need to know whether they follow a normal distribution or not. To do so, we apply the Shappiro-Wilks test.

```{r}
HG <- read.csv(file.choose(), header=T, sep="\t")
HG <- HG[,-4]

WES <- read.csv(file.choose(), header=T, sep="\t")
WES

shapiro_hg <- shapiro.test(HG$Insert.Size)
shapiro_material <- shapiro.test(HG$Material)
shapiro_wes <- shapiro.test(WES$Insert.Size)
```

As they follow a normal distribution, the samples are independent and we have a small sample size (less than 30), we can perform a t-test. However, we must first look at homoscedasticity:

```{r}
var.test(x = HG$Insert.Size,
         y = WES$Insert.Size ) # The test finds no significant differences for variance.
```

Finally, we can now apply the t.test to compare the mean insert size between the two groups of samples.

```{r}
t.test(HG, WES, var.equal=T)
```

# 5. ONCOPRINT

First, we must load the libraries required for this step.

```{r}
library(reshape2) # Flexibly reshape data.
library(ComplexHeatmap) # Provides a highly flexible way to arrange multiple heatmaps and supports various annotation graphics.
```

From the final result obtained in the filtering step, we convert it into a matrix with 0 and 1. Therefore, we are transforming an input list with two columns (one with the sample and the other with the mutated gene) into a matrix with 0 for the WT cases and 1 for the mutated ones.

```{r}
# First we select the dataset with the potential driver mutations post-filtering and select the Sample ID column and the Gene name column.

D <- read.csv(file = "FINAL_POTENTIALDRIVER_MUTATIONS.csv")
head(D)

driver_matrix <- D[, c("sampleNames", "Gene_name")]
driver_matrix

write.csv2(driver_matrix, file = "driver_matrix.csv", row.names = FALSE)
D <- distinct(D)

# We can then transform the list into a matrix of 0 and 1.
file <- xtabs( ~ ., data = D)
write.table(file,"MATRIX_01_DRIVER_MUTATIONS.txt",sep="\t",row.names=TRUE, quote=F)
```
The following code is used to perform an Oncoprint with the results obtained through the filtering steps.
In order to run this code, the oncoPrint funtion needs a list of matrix. As this analysis is only studying whether the gene has a mutation or not (per each sample), we just need to upload another matrix with 1 for wt and 0 for mut.
In this case, for example, we need 6 matrices, one for MYC status, one for Wild Types (non-mutated genes per sample), one for LymphGen results, one for Classification Learner results, and for the genes

```{r}
setwd("C:/Users/User/Desktop/TFM_RESULTS/FINAL_ONCOPRINT")

mat1 <- read.csv2("WT_ONCOPRINT.csv", sep = ",",  row.names = 1)
mat2 <- read.csv2("MYC_ONCOPRINT.csv", sep = ",",  row.names = 1)
mat3 <- read.csv2("LYMPHGEN_ONCOPRINT.csv", sep = ",",  row.names = 1)
mat4 <- read.csv2("CLASSIFICATION_LEARNER_ONCOPRINT.csv", sep = ",",  row.names = 1)
mat5 <- read.csv2("MUTATED_GENES_HG.csv", sep = ",",  row.names = 1)
mat6 <- read.csv2("OTHER_MUTATED_GENES.csv", sep = ",",  row.names = 1)
```

Then, we put all the matrices together in a single list, called mat_list
```{r}
mat_list = list(WT = as.matrix(mat1), MYC = as.matrix(mat2), LYMPHGEN = as.matrix(mat3), CLASSIFICATION = as.matrix(mat4), MUTATED = as.matrix(mat5), OTHER = as.matrix(mat6))
```
The next step is to carry out the oncoprint. Thus, we have selected a colour for each case, the subsections have been determined according to the sections and the total frequency of both mutated genes and samples is shown. For the realisation of this oncoprint, the instructions given at https://jokergoo.github.io/ComplexHeatmap-reference/book/oncoprint.html have been followed.

```{r}
column_title = "OncoPrint in HGBCL, NOS samples."

col = c(MUTATED = "#3FABB4", OTHER = "#176A71", WT = "lightgray", LYMPHGEN= "#D7A424", CLASSIFICATION = "#69B658", MYC = "#B65868" )

row.subsections <- c(1, 1, 1, 9, 22)

hc_list = oncoPrint(mat_list, alter_fun_is_vectorized = FALSE, row_names_side = "left", pct_side = FALSE,  column_title = column_title,
                    alter_fun = list(
                      background = function(...) NULL,
                      MYC = function(x, y, w, h) grid.rect(x, y, w*0.9, h*0.9, 
                                                          gp = gpar(fill = col["MYC"], col = NA)),
                      LYMPHGEN = function(x, y, w, h) grid.rect(x, y, w*0.9, h*0.9, 
                                                                    gp = gpar(fill = col["LYMPHGEN"], col = NA)),
                      CLASSIFICATION = function(x, y, w, h) grid.rect(x, y, w*0.9, h*0.9, 
                                                                gp = gpar(fill = col["CLASSIFICATION"], col = NA)),
                      MUTATED = function(x, y, w, h) grid.rect(x, y, w*0.9, h*0.9, 
                                                                gp = gpar(fill = col["MUTATED"], col = NA)),
                      OTHER = function(x, y, w, h) grid.rect(x, y, w*0.9, h*0.9, 
                                                               gp = gpar(fill = col["OTHER"], col = NA)),
                      WT = function(x, y, w, h) grid.rect(x, y, w*0.9, h*0.9, 
                                                                 gp = gpar(fill = col["WT"], col = NA))
                    ), col = col, show_column_names = TRUE, row_names_gp = gpar(fontsize = 8), row_split = data.frame(rep(c("MUTATED", "OTHER","MYC","CLASSIFICATION", "LYMPHGEN"), row.subsections)))
draw(hc_list)
```
Now, it is very important to analyse the Oncoprint and decide which order of the samples or genes we are most interested in and redo it.

# 6. ENRICHMENT ANALYSIS

First, we must load the libraries required for this enrichment analysis.

```{r}
library(org.Hs.eg.db) #  Genome wide annotation for Human. 
library(clusterProfiler) # R package for comparing biological themes among gene clusters.
library(biomaRt) #  interface to a growing collection of databases implementing the BioMart software suite ( ). 
library(GOstats) # Tools for manipulating GO and microarrays.
library(KEGGREST) # Provides a client interface to the Kyoto Encyclopedia of Genes and Genomes (KEGG) REST server
```

Next, we need to select the list of genes that present potential driver mutations. With this, and by selecting the gene symbol, we can transform it into an ENSEMBL ID and create a list of genes of interest.

```{r}
DRIVER_MUTATIONS <- read.csv2("GENES_PREDICTED.csv", header = F, sep = ",", stringsAsFactors = F)

colnames(DRIVER_MUTATIONS) <- "GENE_SYMBOL"

DRIVER_MUTATIONS <- bitr(DRIVER_MUTATIONS$GENE_SYMBOL, fromType="SYMBOL", toType = "ENSEMBL", OrgDb = "org.Hs.eg.db")

DRIVER_MUTATIONS <- DRIVER_MUTATIONS$ENSEMBL

DRIVER_MUTATIONS <- unlist(mget(DRIVER_MUTATIONS_HG, envir=org.Hs.egENSEMBL2EG,
                       ifnotfound = NA))
```

Now that we have the list of genes of interest in the correct format, we need to make a gene Universe list for enrichment analysis. It, too, must have the genes annotated as ENSEMBL IDs.

```{r}
mart <- useMart("ensembl", dataset = "hsapiens_gene_ensembl")
genes <- getBM(
  attributes=c("hgnc_symbol","entrezgene_id","chromosome_name","start_position","end_position"),
  mart = mart)

all_genes <- na.omit(genes$hgnc_symbol)

all_genes <-unique(all_genes)
write.table(all_genes, "GENE_UNIVERSE.txt")

data <- read.table(file = "GENE_UNIVERSE.txt", header = FALSE) # Based on the panel used 
geneUniverse <- data$V2
geneUniverse <- bitr(geneUniverse, fromType="SYMBOL", toType = "ENSEMBL", OrgDb = "org.Hs.eg.db")

geneUniverse <- geneUniverse$ENSEMBL

geneUniverse <- unlist(mget(geneUniverse, envir=org.Hs.egENSEMBL2EG,
                            ifnotfound = NA))
```

Now, we have to define the parameters and apply the hyperGTest function. 

```{r}
params <- new("GOHyperGParams", geneIds=DRIVER_MUTATIONS ,
              universeGeneIds=geneUniverse,
              annotation="org.Hs.eg.db", ontology="BP",
              pvalueCutoff=0.05, conditional=FALSE,
              testDirection="over")

#Finally, we have to apply the hyperGTest:
hgOver <- hyperGTest(params)
hgOver <- summary(hgOver)
hgOver 

write.csv2(kegg, "GOstats_.csv")
```
Another example would be with KEGGHyperGParams to see which pathways are enriched based on these genes. 

```{r}
options(clusterProfiler.download.method = "wininet") # This line of code was required in order to run the enrichKEGG analysis.

ans.kegg <- enrichKEGG(gene = DRIVER_MUTATIONS,
                       organism = 'hsa')
tab.kegg <- as.data.frame(ans.kegg)
tab.kegg<- subset(tab.kegg)
tab.kegg <- tab.kegg[, 1:6]
write.csv2(tab.kegg, "KEGG_PREDICTED.csv")
```
Although these packages tell us information about enrichment, it is also interesting to know which genes are involved in which pathways or biological processes. For this reason, we consult the DAVID database website: https://david.ncifcrf.gov/.

```{r}
gda <- read.delim("curated_gene_disease_associations.tsv.gz")
disease2gene <- gda[, c("diseaseId", "geneId")]
disease2name <- gda[, c("diseaseId", "diseaseName")]

ans.dis <- enricher(DRIVER_MUTATIONS, TERM2GENE=disease2gene,
                    TERM2NAME=disease2name, pvalueCutoff = 0.05)

tab.dis <- as.data.frame(ans.dis)

diseases <- tab.dis[,1:6]
```

