---
title: "Preparation_Input_Files_LymphGen"
author: "Ariadna Colmenero Cobo de Guzmán"
output: html_document
---
######
# PREPARATION OF THE INPUT FILES FOR LYMPHGEN
######

In order to be able to run **LymphGen** via https://llmpp.nih.gov/lymphgen/index.php, we have to prepare some input files. In this instance, we are using the results, once filtered and correctly predicted, from SNPEff/SNPSnif. To see further information read the **LymphGenInstructions** available at its web page.

First, we must load all the libraries that will be required in order to apply the following code. You can see the versions used in the report.

```{r}
library(org.Hs.eg.db) # Genome wide annotation for Human, primarily based on mapping using Entrez Gene identifiers. 
library(dplyr) # A fast, consistent tool for working with data frame like objects, both in memory and out of memory.
# BiocManager::install("Homo.sapiens")
library(Homo.sapiens) # Contains the Homo.sapiens object to access data from several related annotation packages.
library(GenomicRanges) # Defines general purpose containers for storing and manipulating genomic intervals and variables defined along a genome.
library(biomaRt) # Enables a wide range of powerful online queries from gene annotation to database mining.
library(data.table) # Provides a high-performance version of base R's data.frame.
```
We can then elaborate each of the files necessary for the prediction.

## 1. Sample Annotation file:

This tab-delimited text file contains information about the data  availability of the samples being analysed. Each row indicates a sample to be analyzed. It consists of the following four required columnsin order:

    1) **Sample name**
    2) **Copy number availability**
    3) **BCL2 Translocation**
    4) **BCL6 Translocation**

```{r}
db <- read.table("translocations.csv", sep = ",", stringsAsFactors = F, header=TRUE) # Load the required CSV.
head(db) # We can see the first 6 lines.

SAM <- data.frame(cbind(db$CASE, 0, db$BCL2, db$BCL6), stringsAsFactors = F) # We bind those required columns. 
colnames(SAM) <- c("Sample.ID", "Copy.Number", "BCL2.transloc", "BCL6.transloc") # Give them the proper name indicated by Lymphgen.

head(SAM) #Now, we have created the first needed file. We have to proceed with:
write.table(SAM, "SAMPLE_ANNOTAION.txt", sep="\t") # This is the firt file needed for the prediction.
```
You can see an example of an Annotation File at: https://github.com/ariadnacolmeneroc/Final-Master-Thesis-Ariadna-Colmenero/tree/Supplemental-Information.

## 2. Mutation Flat file:

This is a tab-delimited text file with each row indicating a single alteration on a single sample. If a sample has multiple alterations for a gene, all should be listed on separate rows. First, we need a file which contains the **Sample ID**, the **Gene_name**, the **annotation** and de **Location** (position).

```{r}
muts <- read.csv2("FINAL_DRIVER_MUTATIONS.csv", sep = ",", stringsAsFactors = F) # List of final driver mutations (once they have been filtered).
head(muts) # We check the first six rows. 
 
flat <- muts[, c("CASE", "GENE_NAME", "ANNOTATION", "POSITION")] # From the final driver mutations list, we subset only those columns we are interested in.
flat <- as.data.frame(flat) # We save it as a data.frame.
```
Now we have to transform all that gene symbols to **ENTREZ ID**. For that, we proceed:

```{r}
Gene_names <- flat$GENE_NAME # We only select the column in which there are the gene names. 
flat$GENE_NAME <- mapIds(org.Hs.eg.db, keys = Gene_names, keytype = "SYMBOL", column = "ENTREZID") # We change them for the corresponding ENTREZ ID.
View(flat) # We check whether it has been performed correctly.
write.csv2(muts, "MUTATIONS_ENTREZID_PREDICTION.csv") # We save this object into a csv.
```
Now we can proceed with the selection of the type of mutation. Based on the information given by LymphGen, we know that: 

- **TRUNC**: indicating a nonsense mutation in the coding region of the gene.
- **MUTATION**: indicating a missense or frameshift mutation in the coding region of the gene.
- **Synon**: indicating a mutation in the 5’UTR of the gene or a synonymous mutation.
- 
For that reason, and checking which are the mutations in the file containin the potential driver mutations, we can see which of them are truncating:

```{r}
TRUNC <- c("stopgain", "frameshift deletion") # We select as truncating those mutaions in the potential driver mutations list. 
TRUNC_ROW <- "TRUNC"
flat$Exonic_function[muts$Exonic_function %in% TRUNC] <- "TRUNC" 
subset(flat, !(Exonic_function %in% TRUNC_ROW)) # Subset of all those mutations which are not truncations. 
flat$Exonic_function[!(flat$Exonic_function %in% TRUNC_ROW)] <- "MUTATION" # Give the name Mutation to the rest which are not truncations.

write.csv2(flat, "MUTATIONS_ENTREZID_PREDICTION.csv") # We save the file and check that everything is correct.  

# Next, we change the column names as indicated in the Lymphgen instructions: 
colnames(flat) <- c("Sample", "ENTREZ.ID", "Type", "Location")

write.table(flat[,c("Sample", "ENTREZ.ID", "Type", "Location")], "MutationFlat-PREDICTION.txt", sep = "\t", quote = F, col.names = T, row.names = F) # We save the file in order to use it for the prediction.

# Prior to continue, we can access the different columns to check that we have perfom the subset correctly.
table(flatt$Type)
table(flat$Sample %in% SAM$Sample.ID)
SAM$Sample.ID[!SAM$Sample.ID %in% flat$Sample]
```

## 3. GENE LIST

The **gene list** text file with a single column indicating the NCBI Entrez Gene IDs, for all genes that were analyzed for mutations. Every ENTREZ.ID included in the mutation flat file  should be included on this list, but this list may include genes that were not found to have alternations in any of the samples. This list should apply to all samples run on the array. If the different samples have different gene lists(for example, if they are analyzed on different targeted arrays),then they should be run through the LymphGen algorithm in separate batches.

```{r}
db <- read.csv(file.choose(), header=T, sep="\t") # We select the genes in which there are potential driver mutaitons. 
db

entrez <- data.frame(cbind(mapIds(org.Hs.eg.db, db$Gene, 'ENTREZID', 'SYMBOL')), stringsAsFactors = F) # From the symbol list, we have to transform them into ENTREZ IDs. 
# Next, we apply a set of steps to correctly elaborate the data frame.
entrez <- cbind(rownames(entrez), entrez)
colnames(entrez) <- c("Gene", "ENTREZ.ID")
rownames(entrez) <- NULL
head(entrez)
write.table(data.frame(entrez$ENTREZ.ID), "MutGeneList_168.txt", sep = "\t", quote = F, col.names = T, row.names = F)

# Finally, we remove all NA values and convert it into a list so that we can save it and use it as a gene list.
entrez_flat <- as.list(na.omit(flat$ENTREZ.ID))
entrez_lymphgen <- as.list(entrez$ENTREZ.ID)

a <- as.matrix(append(entrez_flat, entrez_lymphgen))
colnames(a) <- "ENTREZ.ID"
rownames(a) <- NULL
write.table(a, "gene-list-lymph-flat.txt", sep= "\t")
```

## 3.1. OBTAINING ALL GENE'S ENTREZID FOR WES ANALYSIS 

For the cases of the samples analysed with WES, as gene list, all the ENTREZ IDs have been used, although, in the end, we saw that we only needed the list of 114 genes with which the predictor was made. So, this step was repeated with both files to see that the result was the same.

```{r}
mart <- useMart("ensembl", dataset = "hsapiens_gene_ensembl") # Accounting for the information for genes in Homo sapiens.
genes <- getBM(
  attributes=c("hgnc_symbol","entrezgene_id","chromosome_name","start_position","end_position"),
  mart = mart) # Selecting the columns we are interested in.

entrez <- na.omit(genes$entrezgene_id) # Omit all NAs which are in the column belonging to ENTREZ IDs. 
genelist_all_entrezid <-unique(entrez) # We remove all those which are duplicated.
write.table(genelist_all_entrezid, "genelist_all_entrezid.txt") # We save it into a TXT to use it in the predictor.
```

## 4. COPY NUMBER INFORMATION AVAILABLE FOR THE 17 HGBCL, NOS 

For HGBCL, NOS information was available for the CN. Therefore, the following code refers to the creation of the necessary files for the predictor concerning the CN.

## 4.1. CNA GENE LIST
4
First we create a gene list, as we have done before. If this step has already been carried out, do not repeat it:

```{r}
k <- keys(org.Hs.eg.db,keytype="SYMBOL")
entrez <- data.frame(cbind(mapIds(org.Hs.eg.db, k, 'ENTREZID', 'SYMBOL')), stringsAsFactors = F)
entrez <- cbind(rownames(entrez), entrez)
colnames(entrez) <- c("Gene", "ENTREZ.ID")
rownames(entrez) <- NULL
head(entrez)

write.table(data.frame(ENTREZ.ID=entrez[,2]), "CopyGeneList.txt", sep = "\t", quote = F, col.names = T, row.names = F)
```

## 4.2. CNA FLAT FILE

To make the **copy number flat file**, we need an input file containing the following columns referring to the CN event: **Chromosome**, **Start position**, **End position**, **Length**, **Event**. To do so, 

```{r}
CNA <- read.csv(file.choose(), header=T, sep="\t") # We select the file where we have the information for that columns. 

CNA <- CNA[!CNA$Event == c("LOH"),] # We don't take into account the Loss of Heterozigozity (LOH).
CNA <- CNA[!CNA$Event == c("CNN-LOH"),] # Another event we have to remove as it is a LOH.
CNA <- CNA[CNA$Length < 30000000,] # We only select those events which have a length lower than 30000000 bp.

genes <- as.data.frame(org.Hs.egSYMBOL) # We retrieve the gene's symbols. 
num <- as.numeric(nrow(CNA)) 

# We add the same steps as we have just done, so, they are commented. (But, if you only need the code to do this part it is mandatory to run it!)
# entrez <- data.frame(cbind(mapIds(org.Hs.eg.db, genes$symbol, 'ENTREZID', 'SYMBOL')), stringsAsFactors = F)
# entrez <- cbind(rownames(entrez), entrez)
# colnames(entrez) <- c("Gene", "ENTREZ.ID")
# rownames(entrez) <- NULL
# head(entrez)
```
The next step is to treat the dataframe as a genomic range. That is to say, each range has an Rle column referring to the chromosome, an IRange, with the genomic position of the event and, in addition, the metadata columns which include information, for example whether the event is an amplification, a loss of heterozygosity,...

```{r}
AllCNA <- lapply(c(1:nrow(CNA)), function(x){
  coor <- makeGRangesFromDataFrame(CNA[x,c(2:5)], na.rm=TRUE) # Select all columns except the Sample Id to compile the genomic range.
  df <- as.data.frame(subsetByOverlaps(genes(TxDb.Hsapiens.UCSC.hg19.knownGene), coor)) # Select the genes belonging to the coordinates found in the coor object.
  gn <- as.vector(df[,"gene_id"])
  symb <- lapply(gn, function(y){
    if(length(gn)==0){ next }
    else {
      gnsymb <- as.vector(genes[genes$gene_id == y, ]$symbol)
      return(gnsymb) # Return the gene symbol.
    }
  })
  symb <- unlist(symb)
  case <- rep(CNA[x,1],length(symb)) 
  alt <- rep(CNA[x,6],length(symb))
  MatCNA <- as.data.frame(cbind(case, symb, alt)) # Create the required matrix with the case, the gene symbol and the dalteration.
  return(MatCNA)
  print(num-x)
})  

# Then efficiently link multiple dataframes per row.
CNAlist <- bind_rows(AllCNA, .id = "column_label")
CNAlist$ENTREZ.ID <- entrez$ENTREZ.ID[match(CNAlist$symb, entrez$Gene)]
length(CNAlist)
CNAlist <- CNAlist[,c(2,5,4)] 
```


```{r}
unique(CNAlist$alt)

Gain <- c("CN Gain", "Gain")
HETLOSS <- c("CN Loss", "Loss")
HOMDEL <- c("Homozygous Copy Loss", "Homozygous deletion")
Amp <- c("High gain", "High Copy Gain")

Gain_ROW <- "Gain"
CNAlist$alt[CNAlist$alt %in% Gain] <- "Gain"
subset(CNAlist, !(alt %in% Gain_ROW))

HETLOSS_ROW <- "HETLOSS"
CNAlist$alt[CNAlist$alt %in% HETLOSS] <- "HETLOSS"
subset(CNAlist, !(alt %in% HETLOSS_ROW))

HOMDEL_ROW <- "HOMDEL"
CNAlist$alt[CNAlist$alt %in% HOMDEL] <- "HOMDEL"
subset(CNAlist, !(alt %in% HOMDEL_ROW))

Amp_ROW <- "Amp"
CNAlist$alt[CNAlist$alt %in% Amp] <- "Amp"
subset(CNAlist, !(alt %in% Amp_ROW))

unique(CNAlist$alt)


write.csv(CNAlist, file = "CopyFlatList-HG_LAST.txt")
write.csv(entrez$ENTREZ.ID, file = "CopyGeneList-HG_LAST.txt")
```

At this point we already have all the input files that can be loaded into Lymphgen tool: https://llmpp.nih.gov/lymphgen/index.php
